src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -757,11 +757,30 @@ private static bool IsArrayCompatibleReadOnlyInterface(Type type)
             Type type)
         {
             var allProperties = new List<PropertyInfo>();
+            HashSet<string>? propertyNames = null;
 
             Type? baseType = type;
             do
             {
-                allProperties.AddRange(baseType!.GetProperties(DeclaredOnlyLookup));
+                PropertyInfo[] baseTypeProperties = baseType!.GetProperties(DeclaredOnlyLookup);
```

Comment:
Can't we just use something like:

`type.GetProperties(BindingFlags.Instance | BindingFlags.Public)`

?

Reply:
This breaks some unit tests :

```
[xUnit.net 00:01:15.23]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanReadStaticProperty [FAIL]
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanReadStaticProperty [27 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanReadStaticProperty() in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 964
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
[xUnit.net 00:01:15.27]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanReadStaticProperty [FAIL]
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanReadStaticProperty [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: other stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanReadStaticProperty() in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 981
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
[xUnit.net 00:01:15.29]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "InternalProperty") [FAIL]
[xUnit.net 00:01:15.29]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "PrivateProperty") [FAIL]
[xUnit.net 00:01:15.29]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "ProtectedProperty") [FAIL]
[xUnit.net 00:01:15.29]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "ProtectedPrivateSet") [FAIL]
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "InternalProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1113
   at InvokeStub_ConfigurationBinderTests.BindCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "PrivateProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1113
   at InvokeStub_ConfigurationBinderTests.BindCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "ProtectedProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1113
   at InvokeStub_ConfigurationBinderTests.BindCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "ProtectedPrivateSet") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1113
   at InvokeStub_ConfigurationBinderTests.BindCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
[xUnit.net 00:01:15.30]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "InternalProperty") [FAIL]
[xUnit.net 00:01:15.30]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "PrivateProperty") [FAIL]
[xUnit.net 00:01:15.30]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "ProtectedProperty") [FAIL]
[xUnit.net 00:01:15.30]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "ProtectedPrivateSet") [FAIL]
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "InternalProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1047
   at InvokeStub_ConfigurationBinderTests.GetCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "PrivateProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1047
   at InvokeStub_ConfigurationBinderTests.GetCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "ProtectedProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1047
   at InvokeStub_ConfigurationBinderTests.GetCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "ProtectedPrivateSet") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1047
   at InvokeStub_ConfigurationBinderTests.GetCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

Reply:
Yes it works now. Just to be sure, changing only binding flags does not deduplicate virtual fields, we still need the hashset

Reply:
I was not clear enough on my previous comments. What I meant was:

```C#
        private static List<PropertyInfo> GetAllProperties([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type type) =>
            type.GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
```

Reply:
my bad it was obvious :dizzy_face: